### DISCUSSION TITLE
Throw ParameterException from subcommand defined by method

### DISCUSSION THREAD

**Discussion by bartweber** [remkop/picocli#1668] [Q&A]
A subcommand defined by a method needs to throw a `ParameterException`. The `ParameterException` expects `CommandLine` as the first argument. This object `CommandLine` can be obtained from `CommandSpec`. `CommandSpec` can be injected by PicoCLI by annotating a field for this with `@Spec`

```
@Spec
private CommandSpec spec;
```

Using this injected `CommandSpec` to obtain the `CommandLine` will not be the `CommandLine` of the subcommand but of the parent.

**How to obtain the `CommandLine` of the subcommand defined by a method?**

One solution is like:
```
spec.subcommands().get("my-subcommand")
```
But this could cause runtime errors if the command name was changed by a developer which forgot to update the key as well for looking up the `CommandLine` object for the subcommand as is done in this code block.

Also note that '@CommandLine.Spec' is not applicable to parameter of method defining a subcommand.

Another solution is to define the subcommand in a class instead of a method. A disadvantage of this solution is that this will result in a lot of classes for defining simple subcommands.

Is there a better solution for this?

PioCLI version: 4.6.3

**Comment by remkop** [comment]
Hi @bartweber, yes your analysis is correct. 

One idea, like you suggest, is to modify picocli to allow a `@Spec CommandSpec` parameter to methods that define a subcommand. That may be the cleanest solution. Would you be able to provide a pull request for this? 

Another idea (that works with the current version of picocli) is to combat the fragility by using a String constant, something like this:

```java
class MyApp {
    private static final String SUB = "mysubcmd";
    private @Spec CommandSpec spec;

    @Command(name = SUB)
    void sub(@Option(names = "-x") int x) {
        CommandSpec subSpec = spec.subcommands().get(SUB);
        throw new ParameterException(subSpec.commandLine(), "msg");
    }
}
```

**Reply by bartweber** [reply]
Hoi @remkop,

Thank you for you time answering.
I would like to provide a pull request for the idea to allow a `@Spec CommandSpec` parameter to methods that define a subcommand. I didn't work on this project before so I might need some guidance. What is the best way to reach you (or any team members) for any questions about contributing to the project? I've already found this page: https://github.com/remkop/picocli/blob/main/CONTRIBUTING.md.

**Reply by remkop** [reply]
@bartweber I created this ticket: https://github.com/remkop/picocli/issues/1670

Let's continue the discussion about implementation on that ticket.
